### PR TITLE
Remove the token cache entry if GitHub fails

### LINKED ISSUE #148: Authentication token caching
Hi! I've enabled bors-ng on intellij-rust today, and while it have merged a couple of pull requests successfully, it seem to have missed the `r+` on this one:

https://github.com/intellij-rust/intellij-rust/pull/1130

even robots ingnore me

### LINKED ISSUE #158: PR stuck in "waiting in queue" state
Hi! Looks like something's wrong with https://github.com/intellij-rust/intellij-rust/pull/1139.

Status says "bors — Waiting in queue", but no travis builds are in progress.

### PR DESCRIPTION
**Pull request by notriddle**
Fixes #158
Fixes #148

### CODE REVIEW

File: apps/bors_github/lib/github/server.ex [hunk 1]
```diff
@@ -390,6 +390,31 @@ defmodule BorsNG.GitHub.Server do
   end
 
   @doc """
+  Uses a token from the cache, or, if the request fails,
+  retry without using the cached token.
+  """
+  @spec use_token!(ttoken, ttokenreg, ((ttoken) -> term)) ::
+    {:reply, term, ttokenreg}
```
Comment (Macarse) [inline apps/bors_github/lib/github/server.ex]: I didn't know about `@spec`. Did you know it is used by https://github.com/jeremyjh/dialyxir ?

File: apps/bors_github/lib/github/server.ex [hunk 2]
```diff
@@ -390,6 +390,31 @@ defmodule BorsNG.GitHub.Server do
...
+  def use_token!({:installation, installation_xref} = token, state, fun) do
```
Comment (Macarse) [inline apps/bors_github/lib/github/server.ex]: I see the method is defined with the `!` at the end so I traced the calls to find what raises the exception.
It's `HTTPoison.post!` right?

**Comment by Macarse** [comment]
👍

File: apps/bors_github/lib/github/server.ex (hunk 2 above)
Comment (notriddle) [inline apps/bors_github/lib/github/server.ex]: Yep.

File: apps/bors_github/lib/github/server.ex (hunk 1 above)
Comment (notriddle) [inline apps/bors_github/lib/github/server.ex]: Yeah, it's used by dialyxir. That's one reason for having it; bors gates on dialyzer raising no warnings.

**Comment by notriddle** [comment]
bors r=Macarse

File: apps/bors_github/lib/github/server.ex (hunk 1 above)
Comment (Macarse) [inline apps/bors_github/lib/github/server.ex]: where? I don't see dialyxir as a dep

**Pull request merged**